File: sdk/demo/python/src/prohand_demo/connect.py
#!/usr/bin/env python3
"""
ProHand SDK FFI Demo: Connection Test

Tests basic connection to the ProHand IPC host.
"""

import sys
import argparse
from .utils import DemoBase


class ConnectDemo(DemoBase):
    """Connection test demo."""
    
    def __init__(self):
        super().__init__("ProHand IPC Connection Test")
    
    def run(self, command_endpoint: str, status_endpoint: str, hand_streaming_endpoint: str, wrist_streaming_endpoint: str) -> int:
        """Run the connection test."""
        self.banner()
        
        print("\nConnection parameters:")
        print(f"  Command endpoint:       {command_endpoint}")
        print(f"  Status endpoint:         {status_endpoint}")
        print(f"  Hand streaming endpoint: {hand_streaming_endpoint}")
        print(f"  Wrist streaming endpoint: {wrist_streaming_endpoint}")
        
        try:
            self.section("Connecting to IPC host...")
            client = self.sdk.ProHandClient(command_endpoint, status_endpoint, hand_streaming_endpoint, wrist_streaming_endpoint)
            
            # Wait for connection to establish (asynchronous background connection)
            self.info("Waiting for connection to establish...")
            import time
            max_wait = 2.0  # Maximum wait time in seconds
            poll_interval = 0.1  # Poll every 100ms
            start_time = time.time()
            
            while not client.is_connected() and (time.time() - start_time) < max_wait:
                time.sleep(poll_interval)
            
            if not client.is_connected():
                self.error("Failed to establish connection within timeout")
                self.info("Make sure the IPC host is running")
                client.close()
                return 1
            
            self.success("Successfully connected to IPC host")
            
            self.section("Testing communication...")
            client.send_ping()
            time.sleep(0.2)  # Wait for ping response
            self.success("Ping successful!")
            
            self.section("SDK Information:")
            print(f"  Version: {self.sdk.get_version()}")
            
            client.close()
            
            self.success("Connection test completed successfully!")
            return 0
            
        except self.sdk.ConnectionError as e:
            self.error(f"Connection failed: {e}")
            print("\nMake sure the ProHand IPC host is running.")
            return 1
        except self.sdk.ProHandError as e:
            self.error(f"SDK error: {e}")
            return 1
        except Exception as e:
            self.error(f"Unexpected error: {e}")
            import traceback
            traceback.print_exc()
            return 1


def main() -> int:
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="Test connection to ProHand IPC host",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--command-endpoint",
        type=str,
        default="ipc:///tmp/prohand-commands.ipc",
        help="ZMQ command endpoint"
    )
    parser.add_argument(
        "--status-endpoint",
        type=str,
        default="ipc:///tmp/prohand-status.ipc",
        help="ZMQ status endpoint"
    )
    parser.add_argument(
        "--hand-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-hand-streaming.ipc",
        help="ZMQ hand streaming endpoint"
    )
    parser.add_argument(
        "--wrist-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-wrist-streaming.ipc",
        help="ZMQ wrist streaming endpoint"
    )
    args = parser.parse_args()
    
    demo = ConnectDemo()
    return demo.run(args.command_endpoint, args.status_endpoint, args.hand_streaming_endpoint, args.wrist_streaming_endpoint)


if __name__ == "__main__":
    sys.exit(main())

File: sdk/demo/cpp/src/connect.cpp
/**
 * ProHand SDK FFI Demo: Basic Connection Test
 */

#include <iostream>
#include <cxxopts.hpp>
#include <chrono>
#include <thread>

#include <prohand_sdk/ProHandClient.hpp>
#include "prohand_demo/Utils.hpp"

using namespace prohand_sdk;
using namespace prohand_demo;  // For utils namespace

int main(int argc, char** argv) {
    cxxopts::Options options("connect", "Test connection to ProHand IPC host");

    options.add_options()("command-endpoint", "ZMQ command endpoint",
                          cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-commands.ipc"))("status-endpoint", "ZMQ status endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-status.ipc"))("hand-streaming-endpoint", "ZMQ hand streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-hand-streaming.ipc"))("wrist-streaming-endpoint", "ZMQ wrist streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-wrist-streaming.ipc"))("h,help", "Print usage");

    auto result = options.parse(argc, argv);
    
    if (result.count("help")) {
        std::cout << options.help() << std::endl;
        return 0;
    }

    auto commandEndpoint = result["command-endpoint"].as<std::string>();
    auto statusEndpoint = result["status-endpoint"].as<std::string>();
    auto handStreamingEndpoint = result["hand-streaming-endpoint"].as<std::string>();
    auto wristStreamingEndpoint = result["wrist-streaming-endpoint"].as<std::string>();

    utils::printBanner("ProHand IPC Connection Test");

    std::cout << "Connection parameters:\n";
    std::cout << "  Command endpoint:       " << commandEndpoint << "\n";
    std::cout << "  Status endpoint:         " << statusEndpoint << "\n";
    std::cout << "  Hand streaming endpoint: " << handStreamingEndpoint << "\n";
    std::cout << "  Wrist streaming endpoint: " << wristStreamingEndpoint << "\n";

    try {
        utils::printSection("Connecting to IPC host...");
        ProHandClient client(commandEndpoint, statusEndpoint, handStreamingEndpoint, wristStreamingEndpoint);

        // Wait for connection to establish (asynchronous background connection)
        utils::printInfo("Waiting for connection to establish...");
        const double maxWait = 2.0;  // Maximum wait time in seconds
        const double pollInterval = 0.1;  // Poll every 100ms
        auto startTime = std::chrono::steady_clock::now();

        while (!client.isConnected() && 
               std::chrono::duration<double>(std::chrono::steady_clock::now() - startTime).count() < maxWait)
        {
            std::this_thread::sleep_for(std::chrono::duration<double>(pollInterval));
        }

        if (!client.isConnected())
        {
            utils::printError("Failed to establish connection within timeout");
            utils::printInfo("Make sure the IPC host is running");
            return 1;
        }

        utils::printSuccess("Successfully connected to IPC host");

        utils::printSection("Testing communication...");
        client.sendPing();
        utils::printSuccess("Ping successful!");

        utils::printSection("SDK Information:");
        std::cout << "  Version: " << ProHandClient::getVersion() << "\n";

        utils::printSuccess("Connection test completed successfully!");
        return 0;
        
    } catch (const SdkException& e) {
        utils::printError(std::string("Connection failed: ") + e.what());
        std::cout << "\nMake sure the ProHand IPC host is running.\n";
        return 1;
    } catch (const std::exception& e) {
        utils::printError(std::string("Unexpected error: ") + e.what());
        return 1;
    }
}
